### PR TITLE
Associations_project

### PR DESCRIPTION
**Pull request by billodiallo**
In this PR we created a simple user system and secret posts in Ruby on Rails.
- Created a user Authentication model
- Create Event 
- Created a Event controller
- Created view pages for event, show & form
- Create migration files with table rules
- Add devise to user for Authentication (sign_in ,sign_up)
- Adjust associations between models
- Create migration files with required foreign keys
- visitor model, view, controller setup
- The associations between Event and User model over middle table

### CODE REVIEW

File: app/controllers/events_controller.rb
Comment (enaburekhan) [inline app/controllers/events_controller.rb]: Kindly list past events and upcoming events using:
def index
  @past_events = Event.previous_events
 @upcoming_events = Event.upcoming_events
end

File: app/views/events/_form.html.erb
Comment (enaburekhan) [inline app/views/events/_form.html.erb:12]: Move this logic to a helper. There should be absolutely no logic in the view. Views are used to render data that are passed to them by the controller, and not for generating it.

File: app/views/events/show.html.erb
Comment (enaburekhan) [inline app/views/events/show.html.erb]: Also, move this logic to a helper.

File: app/views/events/show.html.erb
Comment (enaburekhan) [inline app/views/events/show.html.erb]: Also, move this logic to a helper.

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Changes required:full_moon:
**Hi team!
Great on this project:clap::clap:.**

**Major highlights**

- The login page has authentication with a password. Thank you for adding this extra feature.
- The readme and PR is very descriptive
- The UI is awesome
- Logged-in users can easily create both past and upcoming events.

However, we still need to make a few changes to get this project approved:muscle:.

**To improve**
- [x] App Behavior

- The user's show page displays events the user has created and events that the user is attending to separated in past and upcoming ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)). This feature is currently missing in the app. Kindly add it.

- [x] EventsController

- `index` action

1.  lists past events, i.e. `@past_events = Event.previous_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)). As highligghted in the comment below. Fix this following the given suggestion.

2. lists ucomping events, i.e. `@upcoming_events = Event.upcoming_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)). Also fix this following the given suggestion below.

- [x] UsersController

- Kindly add the user_controller.rb and implement the following features in the `show action`:

1. lists events created by user, i.e. `@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
2. lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
3. lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

- [x] Database

- an index for `creator_id` column is added to the events table. This is currently missing. Kindly add it. Why indices for foreign keys are necessary? In short: Since the foreign key columns will be used often to retrieve related information, it's important to index them (queries on those columns will be faster with indexes). More details: check this [article](
https://semaphoreci.com/blog/2017/05/09/faster-rails-is-your-database-properly-indexed.html) and this [article](https://sqlperformance.com/2012/11/t-sql-queries/benefits-indexing-foreign-keys.).

- [x] View
- As highlighted in the PR below, Move the logic in _form.html.erb and show.html.erb to a helper. There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it. Check this [article](http://codefol.io/posts/Where-Do-I-Put-My-Code/) and this [article](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers) on how to use a helper in rails. See [screen-shot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from pathwright.

Make the above-required changes and when you are ready, request another review.
Do have questions/comments? Kindly drop them :point_down: and ping me on slack @ Eric Enaburekhan to respond.
Happy hacking:blush:
Eric Enaburekhan

**Comment by denisdiaconu** [comment]
Thank you for the review!
We fixed all the problems.

**Review by GzimAsani — CHANGES_REQUESTED** [review]
STATUS: Changes Required :red_circle: 

### Hello team, your project looks almost complete, thanks for doing most of the requirements asked previously. But unfortunately there is one more requirement to implement which was asked previously:

- [ ] You still have logic in your views as the previous reviewer pointed out, please move them to your helpers
- [ OPTIONAL ] Since you are using users/index page to list the users previous and future events, delete the users/show because you are not using it
- [ OPTIONAL ] Add a button which can link the user to their user/index page 


_**If you have any question related to the required changes feel free to comment here, and ping me on slack @GzimAsani**_

HAPPY CODING :computer: :desktop_computer:

**Comment by billodiallo** [comment]
changed requested done.
Kindly check.
Thanks

**Comment by billodiallo** [comment]
Hi TSEs
we are still waiting for APPROVAL and move to the next project since yesterday.

**Comment by billodiallo** [comment]
24h HOURS

**Review by ElijahTobs — APPROVED** [review]
# Status: Approved! :tada: 

Congratulations! @billodiallo and @denisdiaconu 

Good job implementing the change as requested by the previous TSEs.

## You can now merge and submit.

## Happy Coding :computer: 

_Feel free to reach out to me on slack_ **`@Elijah Ayandokun`**
  **Elijah O. Ayandokun**
[Github](https://github.com/ElijahTobs) | [LinkedIn](https://www.linkedin.com/in/ayandokunelijah/) | [Twitter](https://twitter.com/ElijahTobs)

**Pull request merged**
